lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php

```diff
@@ -224,42 +224,35 @@ private function initRelatedCollection($entity, $class, $fieldName, $parentDqlAl
      *
      * @param array  $data     The instance data.
      * @param string $dqlAlias The DQL alias of the entity's class.
+     * @param array  $rawData  Raw data directly from query.
      *
      * @return object The entity.
      *
      * @throws HydrationException
      */
-    private function getEntity(array $data, $dqlAlias)
+    private function getEntity(array $data, $dqlAlias, array $rawData): object
     {
         $className = $this->_rsm->aliasMap[$dqlAlias];
 
         if (isset($this->_rsm->discriminatorColumns[$dqlAlias])) {
             $fieldName = $this->_rsm->discriminatorColumns[$dqlAlias];
 
-            if (! isset($this->_rsm->metaMappings[$fieldName])) {
-                throw HydrationException::missingDiscriminatorMetaMappingColumn($className, $fieldName, $dqlAlias);
+            if (! isset($rawData[$fieldName])) {
```

Comment:
Please explain why this fixes the issue. I don't know the ORM well, and even if I did understand it enough, I would still ask you to explain this in a commit message.

Reply:
![image](https://user-images.githubusercontent.com/3340882/108623348-3f64c200-7447-11eb-9181-9e646a0227aa.png)
I guess this picture will tell everything.

Reply:
> I guess this picture will tell everything.

Sorry, but not really… what's the value of `$fieldName` here? And you talk a lot about the "wrong instance" but never say what the "wrong instance" here… is it an existing object, or a new instance of the wrong class, and if yes, which class? The parent one, another child?

Reply:
`$data` is something middle between data returned by query (`$rawData`) and just before it's mapped to properties.

in `$data` discriminator column is overwritten by real property with same name. I don't know why discriminator is in data at all.

Reply:
> And you talk a lot about the "wrong instance" but never say what the "wrong instance" here… is it an existing object, or a new instance of the wrong class, and if yes, which class? The parent one, another child?

The test shows where which one is wrong instance.

Reply:
```php
$object       = new JoinedTableWithPropertyAsDiscriminatorColumnChild();
$object->type = 'child2';
```
We persist Child as instanceof `Child`

then we retrieve it
```php
$q = $this->_em->createQuery('SELECT o FROM ' . JoinedTableWithPropertyAsDiscriminatorColumnRoot::class . ' o WHERE o.id = :id');
$q->setParameter('id', $child->id);
$object = $q->getSingleResult();
```
Here object is `Child2` instance, NOT `Child` as it's persisted.

Reply:
> The test shows where which one is wrong instance.

Does it? The test should show the right behavior, I wouldn't expect it to mention the wrong instance at all, unless maybe you assert that the returned instance is not the wrong one, rather than asserting that it is the right one…

Reply:
At the time where you took your screenshot above, what was the value of `$fieldName`?

Reply:
> in $data discriminator column is overwritten by real property with same name. I don't know why discriminator is in $data at all. That's why i pass $rawData and get discriminator from there,before it's overwritten.

Ok so to make the fix more clean, we should probably also make sure the discriminator never ends up in `$data` as well. That would make sense, since it is not supposed to be mapped in the object.

Reply:
>  The test should show the right behavior
Exactly and it fails if I remove my fix. 
> Ok so to make the fix more clean, we should probably also make sure the discriminator never ends up in $data as well. That would make sense, since it is not supposed to be mapped in the object.

Exactly.

Reply:
So maybe this should be done here:

https://github.com/doctrine/orm/blob/9836092602e7935966f4e5b0b354b78975568359/lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php#L343

We should make sure not to gather row data from discriminator columns so as not to pollute `$rowData['data']` with it. You should be able to do that based on the contents of this variable IMO:

https://github.com/doctrine/orm/blob/9836092602e7935966f4e5b0b354b78975568359/lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php#L344